### ISSUE TITLE
Questions regarding the code

### ISSUE THREAD

**Issue by talrub** [NicolasZucchet/minimal-LRU#5]
Hi,
3 questions i have regarding the code:

1. In the paper that presented the 'LRU' at section D.2. in the appendix (page 22 in the paper) the authors don't mention a second optimizer 'Adam' as you sued inside 'create_train_state' funtion at train_helpers.py .

2. Regarding the 'linear_warmup' function at 'train_helpers.py' - are you sure this is a linear warmup? According to the authors, we need to start with lr=1e-7 and linearly increase it.
It looks like you decrease the learning rate since you multiply it by  (step + 1) / end_step   which is a fraction ( end_step >> step).

3. In 'train_epch' function at 'train_helpers.py' why do you call 'update_learning_rate_per_step' after every batch and not only after every epoch (common when using CosineAnnealing scheduler) ?

**Comment by NicolasZucchet** [comment]
Hello, sorry for the late reply! Regarding your questions:
1. The second optimizer is there to remove weight decay for recurrent parameters. This choice is for example done in the S5 code base. We found this setting to work in this regime but we didn't ablate it.
2. The learning rate increases as # steps / # total number of warm-up steps, so it does not start at 10^-7 but at 0 and finishes as base_lr after end_step steps.
3. The learning rate is updated after each iteration. The default optax CosineAnnealing scheduler should do the same thing.

**Comment by NicolasZucchet** [comment]
Closing the issue for now
